File: app/src/main/java/com/kota/Telnet/Model/TelnetFrame.java
package com.kota.Telnet.Model;

import androidx.annotation.NonNull;

import com.kota.Telnet.Reference.TelnetAnsiCode;

import java.util.Vector;

public class TelnetFrame {
    public static final int DEFAULT_COLUMN = 80;
    public static final int DEFAULT_ROW = 24;
    public Vector<TelnetRow> rows = new Vector<>();

    protected void finalize() throws Throwable {
        super.finalize();
    }

    public TelnetFrame() {
        initialData(DEFAULT_ROW);
        clear();
    }

    public TelnetFrame(int row) {
        initialData(row);
        clear();
    }

    public TelnetFrame(TelnetFrame aFrame) {
        set(aFrame);
    }

    public void set(TelnetFrame aFrame) {
        if (this.rows.size() != aFrame.getRowSize()) {
            initialData(aFrame.getRowSize());
        }
        for (int i = 0; i < this.rows.size(); i++) {
            this.rows.get(i).set(aFrame.getRow(i));
        }
    }

    public void initialData(int row) {
        this.rows.clear();
        for (int i = 0; i < row; i++) {
            this.rows.add(new TelnetRow());
        }
        clear();
    }

    public void clear() {
        for (int row = 0; row < this.rows.size(); row++) {
            getRow(row).clear();
        }
    }

    public void setPositionBitSpace(int row, int column, byte aBitSpace) {
        getRow(row).bitSpace[column] = aBitSpace;
    }

    public byte getPositionBitSpace(int row, int column) {
        return getRow(row).bitSpace[column];
    }

    public boolean getPositionBlink(int row, int column) {
        return getRow(row).blink[column];
    }

    public int getPositionData(int row, int column) {
        int data = getRow(row).data[column];
        return data & 255;
    }

    public int getPositionTextColor(int row, int column) {
        byte color_index = getRow(row).textColor[column];
        return TelnetAnsiCode.getTextColor(color_index);
    }

    public int getPositionBackgroundColor(int row, int column) {
        byte color_index = getRow(row).backgroundColor[column];
        return TelnetAnsiCode.getBackgroundColor(color_index);
    }

    public void cleanPositionData(int row, int column) {
        getRow(row).cleanColumn(column);
    }

    public void setRow(int index, TelnetRow aRow) {
        if (index >= 0 && index < this.rows.size()) {
            this.rows.set(index, aRow);
        }
    }

    public TelnetRow getRow(int index) {
        return this.rows.get(index);
    }
    public Vector<TelnetRow> getRows() {
        return this.rows;
    }

    public TelnetRow getFirstRow() {
        return this.rows.firstElement();
    }

    public TelnetRow getLatestRow() {
        return this.rows.lastElement();
    }

    public void switchRow(int index, int andIndex) {
        TelnetRow row = this.rows.get(index);
        this.rows.set(index, this.rows.get(andIndex));
        this.rows.set(andIndex, row);
    }

    public int getRowSize() {
        return this.rows.size();
    }

    @NonNull
    public TelnetFrame clone() {
        return new TelnetFrame(this);
    }

    public boolean isEmpty() {
        for (int i = 0; i < this.rows.size(); i++) {
            if (!this.rows.get(i).isEmpty()) {
                return false;
            }
        }
        return true;
    }

    public TelnetRow removeRow(int index) {
        return this.rows.remove(index);
    }

    // 雙字元判斷
    public void reloadSpace() {
        for (int row = 0; row < getRowSize(); row++) {
            int column = 0;
            while (column < DEFAULT_COLUMN) {
                int data = getPositionData(row, column);
                if (data > 127 && column < 79) {
                    boolean text_color_diff = getPositionTextColor(row, column) != getPositionTextColor(row, column + 1);
                    boolean background_color_diff = getPositionBackgroundColor(row, column) != getPositionBackgroundColor(row, column + 1);
                    if (text_color_diff || background_color_diff) {
                        setPositionBitSpace(row, column, (byte) 3);
                        setPositionBitSpace(row, column + 1, (byte) 4);
                    } else {
                        setPositionBitSpace(row, column, (byte) 1);
                        setPositionBitSpace(row, column + 1, (byte) 2);
                    }
                    column++;
                } else {
                    setPositionBitSpace(row, column, (byte) 0);
                }
                column++;
            }
        }
    }

    public void cleanCachedData() {
        for (TelnetRow row : this.rows) {
            row.cleanCachedData();
        }
    }
}

File: app/src/main/java/com/kota/Telnet/Model/TelnetFrame.kt
package com.kota.Telnet.Model

import com.kota.Telnet.Reference.TelnetAnsiCode
import java.util.*

class TelnetFrame {
    
    companion object {
        const val DEFAULT_COLUMN = 80
        const val DEFAULT_ROW = 24
    }
    
    var rows: Vector<TelnetRow> = Vector()
    
    constructor() {
        initialData(DEFAULT_ROW)
        clear()
    }
    
    constructor(row: Int) {
        initialData(row)
        clear()
    }
    
    constructor(frame: TelnetFrame) {
        set(frame)
    }
    
    fun set(frame: TelnetFrame) {
        if (rows.size != frame.getRowSize()) {
            initialData(frame.getRowSize())
        }
        for (i in 0 until rows.size) {
            rows[i].set(frame.getRow(i))
        }
    }
    
    fun initialData(row: Int) {
        rows.clear()
        repeat(row) {
            rows.add(TelnetRow())
        }
        clear()
    }
    
    fun clear() {
        for (row in 0 until rows.size) {
            getRow(row).clear()
        }
    }
    
    fun setPositionBitSpace(row: Int, column: Int, bitSpace: Byte) {
        getRow(row).bitSpace[column] = bitSpace
    }
    
    fun getPositionBitSpace(row: Int, column: Int): Byte {
        return getRow(row).bitSpace[column]
    }
    
    fun getPositionBlink(row: Int, column: Int): Boolean {
        return getRow(row).blink[column]
    }
    
    fun getPositionData(row: Int, column: Int): Int {
        val data = getRow(row).data[column]
        return data.toInt() and 255
    }
    
    fun getPositionTextColor(row: Int, column: Int): Int {
        val colorIndex = getRow(row).textColor[column]
        return TelnetAnsiCode.getTextColor(colorIndex)
    }
    
    fun getPositionBackgroundColor(row: Int, column: Int): Int {
        val colorIndex = getRow(row).backgroundColor[column]
        return TelnetAnsiCode.getBackgroundColor(colorIndex)
    }
    
    fun cleanPositionData(row: Int, column: Int) {
        getRow(row).cleanColumn(column)
    }
    
    fun setRow(index: Int, row: TelnetRow) {
        if (index in 0 until rows.size) {
            rows[index] = row
        }
    }
    
    fun getRow(index: Int): TelnetRow {
        return rows[index]
    }
    
    fun getRows(): Vector<TelnetRow> {
        return rows
    }
    
    fun getFirstRow(): TelnetRow {
        return rows.firstElement()
    }
    
    fun getLatestRow(): TelnetRow {
        return rows.lastElement()
    }
    
    fun switchRow(index: Int, andIndex: Int) {
        val row = rows[index]
        rows[index] = rows[andIndex]
        rows[andIndex] = row
    }
    
    fun getRowSize(): Int {
        return rows.size
    }
    
    // 新增方便的方法
    fun getRowCount(): Int = rows.size
    fun getColumnCount(): Int = DEFAULT_COLUMN
    
    fun getChar(row: Int, column: Int): Char {
        return getPositionData(row, column).toChar()
    }
    
    fun getAttribute(row: Int, column: Int): Int {
        // 組合屬性：文字顏色、背景顏色、閃爍等
        val textColor = getRow(row).textColor[column].toInt()
        val backgroundColor = getRow(row).backgroundColor[column].toInt()
        val blink = if (getRow(row).blink[column]) 0x80 else 0
        val italic = if (getRow(row).italic[column]) 0x40 else 0
        
        return (backgroundColor shl 16) or (textColor shl 8) or blink or italic
    }
    
    public override fun clone(): TelnetFrame {
        return TelnetFrame(this)
    }
    
    fun isEmpty(): Boolean {
        return rows.all { it.isEmpty() }
    }
    
    fun removeRow(index: Int) {
        rows.removeAt(index)
    }
    
    /** 雙字元判斷 */
    fun reloadSpace() {
        for (row in 0 until getRowSize()) {
            var column = 0
            while (column < DEFAULT_COLUMN) {
                val data = getPositionData(row, column)
                if (data > 127 && column < 79) {
                    val textColorDiff = getPositionTextColor(row, column) != getPositionTextColor(row, column + 1)
                    val backgroundColorDiff = getPositionBackgroundColor(row, column) != getPositionBackgroundColor(row, column + 1)
                    
                    if (textColorDiff || backgroundColorDiff) {
                        setPositionBitSpace(row, column, 3)
                        setPositionBitSpace(row, column + 1, 4)
                    } else {
                        setPositionBitSpace(row, column, 1)
                        setPositionBitSpace(row, column + 1, 2)
                    }
                    column++
                } else {
                    setPositionBitSpace(row, column, 0)
                }
                column++
            }
        }
    }
    
    fun cleanCachedData() {
        rows.forEach { it.cleanCachedData() }
    }
}
